### DISCUSSION TITLE
Data API

### DISCUSSION THREAD

**Discussion by dharma66** [open-policy-agent/community#100] [OPA and Rego] [answered]
Hi,
I'm trying to play around with the AP API's and having a spot of bother.
Thanks to some prior help, I get the HTTP API example working.
What I'm trying to do now is, rather than supply the "subordinates" data as part of the policy document, I want to push that data in via the async push API, which I understand to be a PUT to /v1/data/{any path}
I'm using the docker images as per that demo, and whenever I try to PUT directly to the OPA server (8181), I get a 400 with:

{
    "code": "invalid_parameter",
    "message": "all paths owned by bundle \"nginx\""
}

Other requests direct to OPA, such as GET and POST work fine. I suspect I'm doing something daft, but can't figure out what...

**Comment by anderseknert** [comment]
Hi Phil!

When OPA get policies and data from a bundle, that bundle by default "owns" the paths (in this context, called "roots") declared by that bundle. This means that you can't modify the data from any paths provided by the bundle via the REST API. Unless paths are explicitly declared inside of the bundle, the default will be `""`, meaning _all_ paths will be owned by the bundle (and no data can be modified by the REST API).

If you would like to split this apart, and have the policy be provided via a bundle, but the data via the REST API, you can do so by explicitly declare what "roots" are owned by the bundle, and publish your data on a different root. This is done via a `.manifest` file inside of the bundle. See the [bundle file format docs](https://www.openpolicyagent.org/docs/latest/management-bundles/#bundle-file-format) for more information on how to do that.

As an example - if your policy resides under the "policy" root (i.e. your policy is declared with `package policy`), then your `.manifest` file might look like this:

```json
{
  "roots": ["policy"]
}
```
You can now modify any data that _isn't_ under `/v1/data/policy`. Uploading some data like below would now work:

```json
{
  "my-data": {
     "your-data": "here"
  }
}
```
But this would **not** work:
```json
{
  "policy": {
     "your-data": "here"
  }
}
```
Since the `policy` root is owned by the bundle. Does that make sense? :)

**Reply by dharma66** [reply]
Sorry about this, I'm really struggling with the documentation.
I tried using opa build to create a bundle with a .manifest as you describe, but it seems to rename the file to data.json inside the bundle. opa build --help doesn't seem to mention .manifest files, and I can't find any documentation for opa build...

I tried just creating the bundle by hand using tar, but I still get the 400 - "all paths owned by package nginx", and decisions seem to be using the 'roots' from the data.json that was deployed in an earlier attempt (when .manifest was renamed as such by opa build).

Policy decisions now look like this:

{
    "decision_id": "895fcad7-06b0-47af-8046-470cb6a34f0a",
    "result": {
        "httpapi": {
            "authz": {
                "allow": false
            }
        },
        "roots": [
            "policy"
        ]
    }
}

This is after is deployed the hand-build bundle.tar.gz (using the bundle server from the API example).

If I list policies, I get this:

 "result": [
        {
            "id": "nginx/example.rego",
            "raw": "package httpapi.authz\n\ndefault allow = false\n\n# Allow users to get their own salaries.\nallow {\n\tsome username\n\tinput.method == \"GET\"\n\tinput.path = [\"finance\", \"salary\", username]\n\tinput.user == username\n}\n\n# Allow managers to get their subordinates' salaries.\nallow {\n\tsome username\n\tinput.method == \"GET\"\n\tinput.path = [\"finance\", \"salary\", username]\n\tdata.subordinates[input.user][_] == username\n}\n",
            "ast": {
                "package": {
                    "path": [
                        {
                            "type": "var",
                            "value": "data"
                        },
                        {
                            "type": "string",
                            "value": "httpapi"
                        },
                        {
                            "type": "string",
                            "value": "authz"
                        }
                    ]
                },
                "rules": [
                    {
                        "default": true,
                        "head": {
                            "name": "allow",
                            "value": {
                                "type": "boolean",
                                "value": false
                            }
                        },
                        "body": [
                            {
                                "terms": {
                                    "type": "boolean",
                                    "value": true
                                },
                                "index": 0
                            }
                        ]
                    },
                    {
                        "head": {
                            "name": "allow",
                            "value": {
                                "type": "boolean",
                                "value": true
                            }
                        },
                        "body": [
                            {
                                "terms": [
                                    {
                                        "type": "ref",
                                        "value": [
                                            {
                                                "type": "var",
                                                "value": "eq"
                                            }
                                        ]
                                    },
                                    {
                                        "type": "ref",
                                        "value": [
                                            {
                                                "type": "var",
                                                "value": "input"
                                            },
                                            {
                                                "type": "string",
                                                "value": "method"
                                            }
                                        ]
                                    },
                                    {
                                        "type": "string",
                                        "value": "GET"
                                    }
                                ],
                                "index": 0
                            },
                            {
                                "terms": [
                                    {
                                        "type": "ref",
                                        "value": [
                                            {
                                                "type": "var",
                                                "value": "eq"
                                            }
                                        ]
                                    },
                                    {
                                        "type": "ref",
                                        "value": [
                                            {
                                                "type": "var",
                                                "value": "input"
                                            },
                                            {
                                                "type": "string",
                                                "value": "path"
                                            }
                                        ]
                                    },
                                    {
                                        "type": "array",
                                        "value": [
                                            {
                                                "type": "string",
                                                "value": "finance"
                                            },
                                            {
                                                "type": "string",
                                                "value": "salary"
                                            },
                                            {
                                                "type": "var",
                                                "value": "__local0__"
                                            }
                                        ]
                                    }
                                ],
                                "index": 1
                            },
                            {
                                "terms": [
                                    {
                                        "type": "ref",
                                        "value": [
                                            {
                                                "type": "var",
                                                "value": "eq"
                                            }
                                        ]
                                    },
                                    {
                                        "type": "ref",
                                        "value": [
                                            {
                                                "type": "var",
                                                "value": "input"
                                            },
                                            {
                                                "type": "string",
                                                "value": "user"
                                            }
                                        ]
                                    },
                                    {
                                        "type": "var",
                                        "value": "__local0__"
                                    }
                                ],
                                "index": 2
                            }
                        ]
                    },
                    {
                        "head": {
                            "name": "allow",
                            "value": {
                                "type": "boolean",
                                "value": true
                            }
                        },
                        "body": [
                            {
                                "terms": [
                                    {
                                        "type": "ref",
                                        "value": [
                                            {
                                                "type": "var",
                                                "value": "eq"
                                            }
                                        ]
                                    },
                                    {
                                        "type": "ref",
                                        "value": [
                                            {
                                                "type": "var",
                                                "value": "input"
                                            },
                                            {
                                                "type": "string",
                                                "value": "method"
                                            }
                                        ]
                                    },
                                    {
                                        "type": "string",
                                        "value": "GET"
                                    }
                                ],
                                "index": 0
                            },
                            {
                                "terms": [
                                    {
                                        "type": "ref",
                                        "value": [
                                            {
                                                "type": "var",
                                                "value": "eq"
                                            }
                                        ]
                                    },
                                    {
                                        "type": "ref",
                                        "value": [
                                            {
                                                "type": "var",
                                                "value": "input"
                                            },
                                            {
                                                "type": "string",
                                                "value": "path"
                                            }
                                        ]
                                    },
                                    {
                                        "type": "array",
                                        "value": [
                                            {
                                                "type": "string",
                                                "value": "finance"
                                            },
                                            {
                                                "type": "string",
                                                "value": "salary"
                                            },
                                            {
                                                "type": "var",
                                                "value": "__local1__"
                                            }
                                        ]
                                    }
                                ],
                                "index": 1
                            },
                            {
                                "terms": [
                                    {
                                        "type": "ref",
                                        "value": [
                                            {
                                                "type": "var",
                                                "value": "eq"
                                            }
                                        ]
                                    },
                                    {
                                        "type": "var",
                                        "value": "__local2__"
                                    },
                                    {
                                        "type": "ref",
                                        "value": [
                                            {
                                                "type": "var",
                                                "value": "input"
                                            },
                                            {
                                                "type": "string",
                                                "value": "user"
                                            }
                                        ]
                                    }
                                ],
                                "index": 2,
                                "generated": true
                            },
                            {
                                "terms": [
                                    {
                                        "type": "ref",
                                        "value": [
                                            {
                                                "type": "var",
                                                "value": "eq"
                                            }
                                        ]
                                    },
                                    {
                                        "type": "ref",
                                        "value": [
                                            {
                                                "type": "var",
                                                "value": "data"
                                            },
                                            {
                                                "type": "string",
                                                "value": "subordinates"
                                            },
                                            {
                                                "type": "var",
                                                "value": "__local2__"
                                            },
                                            {
                                                "type": "var",
                                                "value": "$0"
                                            }
                                        ]
                                    },
                                    {
                                        "type": "var",
                                        "value": "__local1__"
                                    }
                                ],
                                "index": 3
                            }
                        ]
                    }
                ],
                "comments": [
                    {
                        "Text": "IEFsbG93IHVzZXJzIHRvIGdldCB0aGVpciBvd24gc2FsYXJpZXMu",
                        "Location": {
                            "file": "/example.rego",
                            "row": 5,
                            "col": 1
                        }
                    },
                    {
                        "Text": "IEFsbG93IG1hbmFnZXJzIHRvIGdldCB0aGVpciBzdWJvcmRpbmF0ZXMnIHNhbGFyaWVzLg==",
                        "Location": {
                            "file": "/example.rego",
                            "row": 13,
                            "col": 1
                        }
                    }
                ]
            }
        }
    ]
}

**Comment by anderseknert** [comment, marked as answer]
The output of `opa build --help` could perhaps do a better job of explaining this, and [this ticket](https://github.com/open-policy-agent/opa/issues/3831) should help when implemented, but here's the key paragraph:

> When -b is specified the 'build' command assumes paths refer to existing bundle files
or directories following the bundle structure. If multiple bundles are provided, their
contents are merged. If there are any merge conflicts (e.g., due to conflicting bundle
roots), the command fails. When loading an existing bundle file, the .manifest from
the input bundle will be included in the output bundle.

If you provide the `-b` (or `--bundle`) flag to `opa build`, the directory structure will be assumed to be that of a bundle, and your .manifest file will get included.

**Reply by dharma66** [reply]
Thanks.. I still get an error, and again, can't track down where it's explained in the docs...

I now have this rego doc (from the API example):

package httpapi.authz

default allow = false

# Allow users to get their own salaries.
allow {
  some username
  input.method == "GET"
  input.path = ["finance", "salary", username]
  input.user == username
}

# Allow managers to get their subordinates' salaries.
allow {
  some username
  input.method == "GET"
  input.path = ["finance", "salary", username]
  data.subordinates[input.user][_] == username
}

And this manifest:

{
  "roots": ["policy"]
}

These are co-located in a folder. From that folder, I run 

opa build -b . 

and get:

error: load error: bundle .: manifest roots [policy] do not permit 'package httpapi.authz' in module 'example.rego'

I tried modifying the roots to include the package:

{
  "roots": ["policy", "hrrpapi.authz"]
}

But still get a very similar error:

error: load error: bundle .: manifest roots [policy  httpapi.authz] do not permit 'package httpapi.authz' in module 'example.rego'

**Reply by anderseknert** [reply]
The root in the manifest would need to match the package, i.e. “ httpapi/authz”.

See the bundle docs for an explanation of “roots”.

**Reply by dharma66** [reply]
That's it!
Many thanks for your help, it's really appreciated!

**Reply by anderseknert** [reply]
I'm pleased to hear that Phil! I'll raise/prioritize some of the issues you encountered when working with this. Thanks for bearing with me!
